src/pybind/mgr/crash/module.py

```diff
@@ -0,0 +1,218 @@
+from mgr_module import MgrModule
+import datetime
+import errno
+import json
+
+
+DATEFMT = '%Y-%m-%d %H:%M:%S.%f'
+
+
+class Module(MgrModule):
+
+    def __init__(self, *args, **kwargs):
+        super(Module, self).__init__(*args, **kwargs)
+
+    def handle_command(self, inbuf, command):
+        for cmd in self.COMMANDS:
+            if cmd['cmd'].startswith(command['prefix']):
+                handler = cmd['handler']
+                break
+        if handler is None:
+            return errno.EINVAL, '', 'unknown command %s' % command['prefix']
+
+        return handler(self, command, inbuf)
+
+    @staticmethod
+    def validate_crash_metadata(inbuf):
+        # raise any exceptions to caller
+        metadata = json.loads(inbuf)
+        if 'crash_id' not in metadata:
+            raise AttributeError
+        return metadata
+
+    @staticmethod
+    def time_from_string(timestr):
+        # drop the 'Z' timezone indication, it's always UTC
+        timestr = timestr.rstrip('Z')
+        return datetime.datetime.strptime(timestr, DATEFMT)
+
+    # command handlers
+
+    def do_info(self, cmd, inbuf):
+        crashid = cmd['id']
+        key = 'crash/%s' % crashid
+        val = self.get_store(key)
+        if not val:
+            return errno.EINVAL, '', 'crash info: %s not found' % crashid
+        return 0, val, ''
+
+    def do_post(self, cmd, inbuf):
+        try:
+            metadata = Module.validate_crash_metadata(inbuf)
+        except Exception as e:
+            return errno.EINVAL, '', 'malformed crash metadata: %s' % e
+
+        crashid = metadata['crash_id']
+        key = 'crash/%s' % crashid
+        if self.get_store(key):
+            return errno.EADDRINUSE, '', 'crash id %s already saved' % crashid
+
+        self.set_store(key, inbuf)
+        return 0, '', 'Posted %s' % crashid
+
+    def do_ls(self, cmd, inbuf):
+        keys = []
+        for key in self.get_store_prefix('crash/').iterkeys():
+            keys.append(key.replace('crash/', ''))
+        return 0, '\n'.join(keys), ''
+
+    def do_rm(self, cmd, inbuf):
+        crashid = cmd['id']
+        key = 'crash/%s' % crashid
+        if self.get_store(key) is None:
+            return errno.EINVAL, '', 'crash rm: id %s not present' % crashid
+        self.set_store(key, None)       # removes key
+        return 0, '', 'Removed %s' % crashid
+
+    def do_prune(self, cmd, inbuf):
+        now = datetime.datetime.utcnow()
+
+        removed = list()
+        for key, meta in self.get_store_prefix('crash/').iteritems():
+            meta = json.loads(meta)
+            keep = meta['keep']
+            stamp = Module.time_from_string(meta['timestamp'])
+            keeptime = datetime.timedelta(days=keep)
+            if stamp <= now - keeptime:
```

Comment:
and if stamp > keeptime break

Reply:
I don't think so? the iterator doesn't necessarily return in timestamp order

Reply:
Oh, currently the crash_ids are $date_$uuid so that they sort in time order.  I suppose that could change, but the intent was to make this easy

Reply:
they sort in the filesystem that way, but how do they come out of config-key, and is that guaranteed?  In any event if this optimization really matters you have too many dumps stored